xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -633,24 +732,46 @@ private void handleCdsResponse(DiscoveryResponse cdsResponse) {
         cdsResponse.getVersionInfo());
 
     // Update local CDS cache with data in this response.
+    absentCdsResources.removeAll(clusterUpdates.keySet());
+    for (String clusterName : clusterNamesToClusterUpdates.keySet()) {
+      if (!clusterUpdates.containsKey(clusterName)) {
+        // Some previously existing resource no longer exists.
+        absentCdsResources.add(clusterName);
+      }
+    }
     clusterNamesToClusterUpdates.clear();
     clusterNamesToClusterUpdates.putAll(clusterUpdates);
 
     // Remove EDS cache entries for ClusterLoadAssignments not referenced by this CDS response.
+    for (String clusterName : clusterNamesToEndpointUpdates.keySet()) {
+      if (!edsServices.contains(clusterName)) {
+        absentEdsResources.add(clusterName);
+      }
+    }
```

Comment:
Yes, it's missing `onError()` notification for endpoint watchers that are watching on EDS resources removed in this CDS update. Fixed.

According to design spec:

> In RDS and EDS, when the client is subscribed to multiple resources, the server may send updates for only those resource(s) that have changed.  The absence of a resource in a response does not indicate that the resource was deleted; it indicates that the resource has not changed.  This means that the API does not actually have a way to tell the client that a resource has been deleted.  The expectation is that the client will have some external indication that it no longer needs a resource: __in the case of RDS, it will be an LDS update that removes the reference to a particular RouteConfiguration name, and in the case of EDS, it will be a CDS update that removes the reference to a particular EDS service name.  When that LDS or CDS update is received, the client will unsubscribe from any RDS or EDS resource that is no longer referenced by the LDS or CDS data and delete it from its local cache__.

EDS resource deletion is done by CDS response. This is how the protocol works.